### PR TITLE
Xccl group

### PR DESCRIPTION
**Pull request by Chao1Han**
Fixes #ISSUE_NUMBER

### CODE REVIEW

File: torch/csrc/distributed/c10d/ProcessGroupXCCL.hpp [hunk 1]
```diff
@@ -151,32 +163,60 @@ class TORCH_API ProcessGroupXCCL : public Backend {
       const std::vector<at::Tensor>& inputs = {},
       const std::vector<at::Tensor>& outputs = {});
 
-  template <typename Fn>
+  template <typename Fn, typename input_t, typename output_t>
+  c10::intrusive_ptr<Work> collective(
```
Comment (zhangxiaoli73) [inline torch/csrc/distributed/c10d/ProcessGroupXCCL.hpp:167]: Would you use the new `collective` definition in first PR? We don't want to change basic API in XCCL backend back and forth.

File: torch/csrc/distributed/c10d/ProcessGroupXCCL.cpp [hunk 2]
```diff
@@ -199,13 +256,24 @@ void ProcessGroupXCCL::WorkXCCL::synchronizeInternal(
...
   synchronizeInternal(timeout);
   return true;
 }
 
+constexpr const char* MULTI_DEVICE_ERROR_MSG =
+    "Expecting one tensor only but got multiple. You are probably using multiple "
+    "devices under one thread. The support for such usage has been deprecated. "
+    "For details, please refer to "
+    "https://pytorch.org/docs/stable/distributed.html#multi-gpu-collective-functions. "
+    "ProcessGroupXCCL continues supporting multi-process and multi-thread modes.";
```
Comment (zhangxiaoli73) [inline torch/csrc/distributed/c10d/ProcessGroupXCCL.cpp:275]: We don't support mult-thread modes.

Comment (Chao1Han) [reply torch/csrc/distributed/c10d/ProcessGroupXCCL.cpp]: done

File: torch/csrc/distributed/c10d/ProcessGroupXCCL.hpp (hunk 1 above)
Comment (Chao1Han) [reply torch/csrc/distributed/c10d/ProcessGroupXCCL.hpp]: rm allgatherv and align with nccl

**Pull request closed without merging**